src/coreclr/jit/ifconversion.cpp

```diff
@@ -656,10 +656,11 @@ bool OptIfConversionDsc::optIfConvert()
 
     if (!m_comp->compStressCompile(Compiler::STRESS_IF_CONVERSION_INNER_LOOPS, 25))
     {
-        // Don't optimise the block if it is inside a loop
-        // When inside a loop, branches are quicker than selects.
+        // Don't optimise the block if it is inside a loop. Loop-carried
+        // dependencies can cause significant stalls if if-converted.
         // Detect via the block weight as that will be high when inside a loop.
-        if (m_startBlock->getBBWeight(m_comp) > BB_UNITY_WEIGHT)
+
+        if (m_startBlock->getBBWeight(m_comp) > BB_UNITY_WEIGHT * 1.05)
```

Comment:
Would be nice to stylize these comparisons, so maybe add a new `fgProfileWeight...` helper here?

In particular these sorts of checks are often parts of profitability heuristics which we might want to vary under stress and/or expose to some machine learning mechanism.

Reply:
I don't think playing around with this threshold makes much sense given its use as an "is inside loop" check. The precise check happens below when this check is false.

Reply:
This is not really checking if the block is likely to be inside a loop, it is checking if the block is likely to be inside a frequently executed loop.

If we assume profile data is representative, then perhaps this is fine; if we think it might not be representative, then perhaps we should be checking `bbNatLoopNum` and if set, either bypass the conversion, or go on to compare the weight of this block to the weight of the loop entry (so either "in a loop" or "frequently executed wrt the loop"). I wonder if this would subsume the need for an `optReachable` check.

Do we know if the cases in #82106 did not get handled by our loop recognition? I don't recall why we didn't check this initially -- do we no longer trust the loop table? Seems like it might still be good enough.

Reply:
The regression in #82106 was because we if-converted inside an unnatural loop that was not recognized by our loop recognition. The loop is this code: 
https://github.com/dotnet/runtime/blob/0202b242169f8d5955e0411dc5e4f42a5d016c1b/src/libraries/System.Private.CoreLib/src/System/SpanHelpers.Char.cs#L87-L137

That's why we ended up with `optReachable` here.

There is an alternative to simply get rid of this block weight check.

Reply:
Ok... let's leave this as is for now.

@BruceForstall do we have a work item for loops we ought to recognize but don't? I see #43713 so perhaps that's good enough for now.